### ISSUE TITLE
IMPORTANT: Missing ssl_ca_file in Spilo Config

### ISSUE THREAD

**Issue by MarkCupitt** [timescale/timescaledb-docker-ha#435]
Patroni supports the inclusion of `ssl_ca_file` so that PostgreSQL can use custom tls certs with root cert validation (we require it for teleport db access, our database is airgapped, so we proxy to it via teleport)

`patroni/features/environment.py` supports `ssl_ca_file`
`patroni/patroni/postgresql/validator.py` supports `ssl_ca_file`
`helm-charts/charts/timescaledb-single/values.yaml` supports `ssl_ca_file`

However

`timescaledb-docker-ha/scripts/configure_spilo.py` omits `ssl_ca_file` 

which means that postgres never gets the ca and tls fails for any session that requires tls validation

This makes TimescaleDb unusable in many scenarios

**Comment by MarkCupitt** [comment]
I'm thinking that adding one line may solve this ...

`timescaledb-docker-ha/scripts/configure_spilo.py`

```
    ssl: 'on'
    ssl_cert_file: {{SSL_CERTIFICATE_FILE}}
    ssl_key_file: {{SSL_PRIVATE_KEY_FILE}}
    ssl_ca_file: {{{postgresql.parameters.ssl_ca_file}}}
    shared_preload_libraries: 'bg_mon,pg_stat_statements,pgextwlist,pg_auth_mon'
```
There may need to be some logic somewhere to exclude it if its not supplied, not sure what postgreSql will do if its set to empty

**Comment by MarkCupitt** [comment]
So, I built a custom image after making the above changes using:

`PG_MAJOR="15" PG_VERSIONS="15.5" TIMESCALEDB_VERSIONS="2.13.0" make build`

Tagged it and uploaded it to my Dockerhub, `docker pull markcupitt/timescaledb:v1.0.0`, loaded it to our Kubernetes cluster and can confirm I am now able to connect via SSL as originally expected.

I will next try it by removing the ca file from the pod, to test what PostgreSQL does when no ca is specified, but this ends up in the postgres parameters

```
ssl_ca_file: 
```

**Comment by MarkCupitt** [comment]
I did this test, and can confirm that PostgreSQL threw a fit :) with the following log entry

```
2024-01-10 04:11:52 UTC [32]: [659e1908.20-1] @,app= [F0000] FATAL:  could not load root certificate file "/etc/certificate/ca.crt": SSL error code 2147483650
```

So some logic in there somewhere is obviously required to ensure taht parameter is populated with the correct ca or not included

**Comment by MarkCupitt** [comment]
Some additional issues with the Helm chart can be found at:

https://github.com/timescale/helm-charts/issues/642
https://github.com/timescale/helm-charts/issues/639

**Comment by alexeyklyukin** [comment]
The configure_spilo.py is deprecated long ago. Patroni is configured from the helm chart:

https://github.com/timescale/helm-charts/blob/7ded6b654c956a3f6dc119d90b47a0262eba600e/charts/timescaledb-single/values.yaml#L193

That configuration is stored in its own configmap
https://github.com/timescale/helm-charts/blob/main/charts/timescaledb-single/templates/configmap-patroni.yaml

and is mounted into the Patroni container
https://github.com/timescale/helm-charts/blob/7ded6b654c956a3f6dc119d90b47a0262eba600e/charts/timescaledb-single/templates/statefulset-timescaledb.yaml#L291

using the volume defined from that configmap:
https://github.com/timescale/helm-charts/blob/7ded6b654c956a3f6dc119d90b47a0262eba600e/charts/timescaledb-single/templates/statefulset-timescaledb.yaml#L480

**Comment by alexeyklyukin** [comment]
@MarkCupitt one may need to add support for the `ssl_ca_file` into the helm chart, but `configure_spilo.py` is a wrong place to add it. It is  still there for compatibility reasons with the zalando `postgres-operator`, although that compatibility is going to be removed at some point in the future, since the original spilo image will probably drift away from this one, and Patroni supports new parameters that the `configure_spilo.py` here is not aware of.

I think if one wants to retain this compatibility, the right actions would be moving `conifgure_spilo.py` and  `augment_patroni_configuraiton.py` into the scripts configmap in the helm chart, decoupling it from this docker image.

**Comment by MarkCupitt** [comment]
@alexeyklyukin If we include the ssl_ca_file in the helm chart, which we can in the postgres section, it IS included in the configmap, however it gets stripped out and does not end up in the postgres config in the container, I have verified that. 

If I add the ssl_ca_file into a running container postgres parameters and reload the config, it works as expected.

SO whatever script runs at container init is the point of failure, I presumed it was the configure_spilo.py, in the container as part of the patroni deployment process but based on what you say, there must be some other process that runs????

Do you know what that is, so I can look at a fix, I have only been able to find the configure_spilo script in the container that seems to access that parameter... I did a global repo search

I don't think this is a helm issue as the chart does what it is supposed to and ssl_ca_file does end up in the config map, and is mounted, it breaks sown after that point

**Comment by MarkCupitt** [comment]
I think I have found what is gong on, there is a job that runs and seems to up date the Postgrs config which overwrites any values we supply in the confiog anyway .. I cna see the other bug we found with the 

`"maintenance_work_mem": "64MB",`

Is also set here as well 

```
  /usr/bin/curl --connect-timeout 30 --include --request PATCH --data 
		
		  {
		  "loop_wait": 10,
		  "maximum_lag_on_failover": 33554432,
		  "postgresql": {
			"parameters": {
			  "archive_command": "/etc/timescaledb/scripts/pgbackrest_archive.sh %p",
			  "archive_mode": "on",
			  "archive_timeout": "1800s",
			  "auto_explain.log_min_duration": "10s",
			  "autovacuum_analyze_scale_factor": 0.02,
			  "autovacuum_max_workers": 10,
			  "autovacuum_naptime": "5s",
			  "autovacuum_vacuum_cost_limit": 500,
			  "autovacuum_vacuum_scale_factor": 0.05,
			  "hot_standby": "on",
			  "huge_pages": "off",
			  "log_autovacuum_min_duration": "1min",
			  "log_checkpoints": "on",
			  "log_connections": "on",
			  "log_disconnections": "on",
			  "log_line_prefix": "%t [%p]: [%c-%l] %u@%d,app=%a [%e] ",
			  "log_lock_waits": "on",
			  "log_min_duration_statement": "1s",
			  "log_statement": "ddl",
			  "maintenance_work_mem": "64MB",
			  "max_connections": 200,
			  "max_prepared_transactions": 150,
			  "max_worker_processes": "60",
			  "password_encryption": "scram-sha-256",
			  "pg_stat_statements.max": "10000",
			  "pg_stat_statements.track": "all",
			  "pgaudit.log": "all, -misc",
			  "pgaudit.log_catalog": "off",
			  "pgaudit.log_parameter": "on",
			  "shared_buffers": "1GB",
			  "shared_preload_libraries": "timescaledb,pg_stat_statements",
			  "ssl": "on",
			  "ssl_cert_file": "/etc/certificate/tls.crt",
			  "ssl_key_file": "/etc/certificate/tls.key",
			  "tcp_keepalives_count": "5",
			  "tcp_keepalives_idle": "60",
			  "tcp_keepalives_interval": "60",
			  "temp_file_limit": "1GB",
			  "timescaledb.passfile": "../.pgpass",
			  "unix_socket_directories": "/var/run/postgresql",
			  "unix_socket_permissions": "0750",
			  "wal_level": "hot_standby",
			  "wal_log_hints": "on"
			},
			"use_pg_rewind": true,
			"use_slots": true
		  },
		  "retry_timeout": 10,
		  "ttl": 30
		}
"http://timescaledb-config:8008/config"
exit 0
```

**Comment by alexeyklyukin** [comment]
@MarkCupitt the container init is at
https://github.com/timescale/timescaledb-docker-ha/blob/26d75a1e50e543123c6a692e6551ddf95bed870f/timescaledb_entrypoint.sh#L1-L50

unless you specify the `SPILO_CONFIGURATION` environment variable, which the helm chart doesn't, 
you'll get straight to Patroni configuration defined in the `values.yaml` of the helm chart.

In the helm chart values, you have the the  `certificateSecretName` in the helm chart, allowing your to override the secret for mounting certificates into the container, and `patroni.bootstrap.dcs.postgresql.parameters` where you can reference `ssl_ca_file`, pointing it to the content you've put into the `certificateSecretName` secret.

The default behavior, if you don't specify the custom secret in the `certificateSecretName`, is to write a self-signed one at https://github.com/timescale/helm-charts/blob/7ded6b654c956a3f6dc119d90b47a0262eba600e/charts/timescaledb-single/templates/secret-certificate.yaml

**Comment by MarkCupitt** [comment]
@alexeyklyukin in a Kubernetes environment adding a Ca in certificateSecretName breaks Patroni. This certificate is used to talk to the Kubernetes API and of course. on a custom cert, kubernetes does not know about it. Patroni DCS uses teh Kubernetes API for its object store. Its the first thing we tried.

The self signed cert does happen, but that cert cannot be used with Teleport per our use case we originally stated

ssl_ca_file in postgres parameters is REMOVEd by patroni, I gave you the code where that happened 

This is going in circles .. we will have to find alternatives I guess .. an adjust manage our own builds .. 

Bottom line is that the container needs to be fixed, we proved it and have it running in our own build, I gave the changes we made to get it working .. the only consideration is if a ca file is not specified, then some logic is required to remove the ssl_ca_file form the config or postgres will break, I suspect that people believe that certificateSecretName is the way to do it, its not, it breaks Patroni in a Kubernetes environment as I stated above

**Comment by MarkCupitt** [comment]
@alexeyklyukin our decision at this stage is that until timescaledb can fix the containers so we can specify a ca, we have no choice but to maintain our own builds (Which we are using now) ... We will drop the helm deployment and use a manifest based deployment using KLUCTL as the deployment tool. We will make our code and configs available in our GIthub Repos for anyone else who needs it, once we have it looking respectable

We will happily go back to official builds once the container can be given a Ca in the postgres parms section of the patroni config and it's not removed, unfortunately, we need to progress our project, our timelines are slipping significantly now

**Comment by MarkCupitt** [comment]
@alexeyklyukin if you feel that there is nothing wrong with the container or the patroni script , then please close this ticket, The container is unusable for us the way it is right now .. 

However I really hope you guys will look at this issue, replicate it, and fix it so we can go back to the official builds OR tell us what we need to do to supply our own ca if its is indeed NOT the patroni script in the container that is the root cause of this issue
